### ISSUE TITLE
In the inventory screen, half-stacks of items recombine when dropped onto another item

### ISSUE THREAD

**Issue by unit41** [minetest/minetest#1115]
1. In the inventory screen, right click a stack of items to pick up half of it. As expected, the other half remains in the original slot.
2. Right or left click a slot containing a different item.

-- The split stack recombines into the destination slot. Instead, the half-stack left behind in the original slot should not move.

**Comment by nerzhul** [comment]
is this already present ?

**Comment by RobertZenz** [comment]
Yes, it is still present, I can reproduce it on the latest HEAD.

**Comment by rubenwardy** [comment]
I can reproduce this. Should it just swap the selected and target stacks?

**Comment by rubenwardy** [comment]
This is kind of annoying to fix, as items that are being dragged are actually still in their previous location, and the server needs to know where all stacks are at a time.

By this I mean, say you have half a stack and you then click a slot. There is already a stack at that position, so you can't just place the selection there - you need to move the stack first. The current way is to swap the `from` slot with the `to` slot, then pick up the `from` slot again (which was the `to` slot). This is what causes the stacks to be recombined.

I can't really see a way to fix this without a large rewrite of the inventory list view. I'd rather wait until I get around to refactoring formspecs.

**Comment by raymoo** [comment]
Where does a stack go when you grab an item from the crafting output?

**Comment by rubenwardy** [comment]
It stays in the output until you drop it it some where else

**Comment by HybridDog** [comment]
You could use some empty item slot as buffer.
When putting the half stack to the destination, the item which was there previously becomes moved to the buffer slot, which is simply the next free slot next to the destination slot.

**Comment by raymoo** [comment]
@HybridDog What if your inventory is full?

**Comment by raymoo** [comment]
By the way, I think the way Minecraft does it is by keeping track on the server side what is being held by the inventory cursor. You can sort of tell by the way that lag degrades inventory manipulation in that game (also a reason you would want client prediction if we used that solution).

**Comment by HybridDog** [comment]
> What if your inventory is full?

Then it doesn't work. This may make modders add an empty inventory slot hidden behind a formspec element which is not desirable.

> […] client prediction […]

You could do all inventory movements client-side until you don't hold anything. When you put the item into a slot without taking one (i.e. the slot is empty or the whole hold stack is added to the current stack at the slot) all movements are performed, beginning with the last one, and sent to server,
e.g. when there are the slots a, b and c (c is an empty slot), the client takes half of the stack from slot a, then clicks on slot b, so the half held is swapped with the stack in b, after that clicks on c, now nothing is held, so it sends to the server that the item from b is moved to c, then half of the items from a is moved to b.
There's just one detriment: The server may use the inventory while the client is performing movement client-side, e.g. if you swap the fuel of a furnace, it begins using the new fuel first when you put the old fuel somewhere that you don't hold something.

**Comment by raymoo** [comment]
Why is that a better solution than keeping track of the held item on the server? (Other than not requiring a change on the server)

Some advantages of tracking the held item are:
 - The swap between a and b happens closer to when the player thinks it happens. For example, if you have a non fuel object in the fuel slot of the furnace, and you swap a half stack of coal with it, it's better if it immediately starts burning rather than waiting for the player to put down the original stack.
 - It solves the problem more simply, conceptually, making it easier to code and debug.

I'm also not sure what happens if the swapping loops around. For example, half of a -> b -> c -> remaining half of a -> new position of b -> etc. It's entirely possible that it works out fine, but I'm not sure.

**Comment by HybridDog** [comment]
Tracking the held item should work if it's only done when a high_synchronicity flag is set in the inventory, else you always have to wait for the server to recognize where you clicked and send redundant information about holding and releasing.

> I'm also not sure what happens if the swapping loops around. For example, half of a -> b -> c -> remaining half of a -> new position of b -> etc. It's entirely possible that it works out fine, but I'm not sure.

It works like a LIFO queue.
```
a/2 -> b
b -> c
c -> a
a/2 -> c (this is the new position of b, isn't it)
c -> empty_slot (the end of the etc.)
```
The movements are performed beginning from the last one the player did.
```
c → some_slot  # c empty now
a/2 → c  # c and a are a/2 now
c → a  # since half of a was taken, a is a now, c is empty again
b → c  # b is empty now
a/2 → b  # a and b are a/2 now
```
The player can't cause loops, only circles. If you find and remove the circles before sending to server, less movements are performed server-side. BTW: It's likely more efficient to send all movements at once instead of one after another.

Consider this movement:
a → b
b → b
b → a
You take an item from a, swap it with the item in b, swap it back and then put it back to a. If you remove this circle, you even don't need to send the movement.
If you consider that players don't likely move items in circles, you could simply omit the cycle detection.
Also, there're on_put and on_take functions for inventories. If a mod logs when you moved which item, it doesn't know the movements if the circle is removed.

**Comment by IrrlichtMt4** [comment]
Tracking is great but would increase lag by a lot in servers. That's why a code rewrite in form specs would help or predictions than the one that is being suggested
